### PR TITLE
PS-5117 : Cannot upgrade from PS 5.7 to 8.0 if datadir is bootstrappe…

### PR DESCRIPTION
**Pull request by satya-bodapati**
…d with --innodb-encrypt-tables

Problem:
-------

Upgrade from 5.7 to 8.0 doesn't work in the following scenarios.

1. 5.7 datadir has encrypted tables like mysql.plugin/servers (due to
   --innodb-encrypted-tables=ON during bootstrap)

2. presence of --innodb-encrypt-tables as mysqld option during upgrade
   (ie when PS-8.0 mysqld is started on 5.7 datadir)

Symptoms:
For 1) mysql_upgrade fails at "ALTER TABLE mysql.plugin TABLESPACE=mysql" with
error saying "source is encrypted and Destination tablespace (mysql) is unencrypted

For 2) In Debug builds, there is assertion failure when dd_properties table creation
fails. Error message "mysql tablespace cannot contain encrypted table".

Fix:
----
For 1) Detect during upgrade if table in mysql database in encrypted, and create
mysqld.ibd as encrypted.

For 2) If --innodb-encrypt-tables is present during upgrade or normal bootstrap
encrypt mysql.ibd

Once mysql.ibd is encrypted, next startup requires innodb_encrypt_tables=ON.
Note that it can be turned off later.

### CODE REVIEW

File: mysql-test/t/dd_upgrade_encrypted.test [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# This test contains zipped 5.7 data directory with innodb partitioned tables.
```
Comment (laurynas-biveinis) [inline mysql-test/t/dd_upgrade_encrypted.test]: Probably not "innodb partitioned tables"?

**Review by laurynas-biveinis — COMMENTED** [review]
Perhaps this scenario is better handled by creating mysql tablespace encrypted?

File: mysql-test/t/dd_upgrade_encrypted.test (hunk 1 above)
Comment (satya-bodapati) [reply mysql-test/t/dd_upgrade_encrypted.test]: Ok, this is copy-paste from existing upgrade tests.

And regarding encrypting mysql tablespace, yes that is an option. The error could still happen with --innodb_encrypt_tables=OFF too.

So how do we decide to encrypt mysql tablespace? Also, since this is the main hardcoded DD tabespace, I was thinking there might be issues with encrypting mysql.ibd. (May be leave some pages unencrypted like in innodb_system). Just thinking aloud.

Comment (laurynas-biveinis) [reply mysql-test/t/dd_upgrade_encrypted.test]: A use case is that we have a 5.7 everything-is-encrypted installation and we want to upgrade to 8.0 and maintain the everything-is-encrypted property.

Comment (satya-bodapati) [reply mysql-test/t/dd_upgrade_encrypted.test]: Ok, let me think more on this

Comment (satya-bodapati) [reply mysql-test/t/dd_upgrade_encrypted.test]: solved the issue of innodb_encrypt_tables=OFF during upgrade by detecting if one of tables in mysql databse is encrypted or not.

And implemented encryption of mysql.ibd

File: mysql-test/t/dd_upgrade_encrypted.test [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+call mtr.add_suppression("Failed to create or recover replication info repositories.");
+call mtr.add_suppression("db.opt file not found for test database. Using default Character set");
+call mtr.add_suppression("Skip re-populating collations and character sets tables in InnoDB read-only mode");
+call mtr.add_suppression("Skipped updating resource group metadata in InnoDB read only mode");
+--enable_query_log
+
+--echo # Set different paths for --datadir
+let $MYSQLD_DATADIR1 = $MYSQL_TMP_DIR/data57_encrypted/data;
+
+--echo # Copy the remote tablespace & DB zip files from suite location to working location.
```
Comment (laurynas-biveinis) [inline mysql-test/t/dd_upgrade_encrypted.test]: Not "remote tablespace", the comment looks redundant anyway

File: storage/innobase/handler/ha_innodb.cc [hunk 3]
```diff
@@ -4955,6 +4959,9 @@ static bool dd_open_hardcoded(space_id_t space_id, const char *filename) {
   }
 
   if (fail) {
+    ib::error(ER_XB_MSG_2)
+        << "If mysql.ibd opening failed with flags mismatch, try startup with"
+        << " innodb_encrypt_tables = ON/OFF";
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/ha_innodb.cc]: Can you rephrase, when to try ON and when to try OFF?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/t/dd_upgrade_encrypted.test (hunk 2 above)
Comment (satya-bodapati) [reply mysql-test/t/dd_upgrade_encrypted.test]: came from existring upgrade test, I didn't bother

File: storage/innobase/handler/ha_innodb.cc (hunk 3 above)
Comment (satya-bodapati) [reply storage/innobase/handler/ha_innodb.cc]: Done

**Comment by satya-bodapati** [comment]
Addressed review comments

**Review by laurynas-biveinis — APPROVED** [review]
